File: plugins/plugin-sdk-js/examples/virtual_light.js
#!/usr/bin/env node
'use strict';

/**
 * virtual_light.js — A software-only dimmable light plugin for HomeCore.
 *
 * Simulates a dimmable light bulb.  On startup it registers with the broker,
 * publishes its initial state, then toggles on/off and steps brightness every
 * 5 seconds.  It also responds to commands received via MQTT.
 *
 * Usage:
 *   npm install
 *   node examples/virtual_light.js
 *   node examples/virtual_light.js --broker 192.168.1.10 --port 1883
 *   HC_BROKER_HOST=192.168.1.10 node examples/virtual_light.js
 */

const { PluginBase } = require('../index');

const DEVICE_ID = 'light.virtual_js_01';
const CAPABILITIES = {
  on:         { type: 'boolean' },
  brightness: { type: 'integer', minimum: 0, maximum: 255 },
};

class VirtualLightPlugin extends PluginBase {
  constructor(options = {}) {
    super({ pluginId: options.id || 'plugin.virtual_js', ...options });
    this._state = { on: false, brightness: 128 };
  }

  onConnect() {
    this.registerDevice(DEVICE_ID, 'Virtual Light (Node.js)', CAPABILITIES, 'living_room');
    this.publishAvailability(DEVICE_ID, true);
    this.publishState(DEVICE_ID, { ...this._state });
    this.publishPluginStatus('active');

    console.log(`[${this.pluginId}] Device registered — initial state:`, this._state);

    // Periodic toggle every 5 seconds.
    setInterval(() => {
      this._state.on         = !this._state.on;
      this._state.brightness = (this._state.brightness + 16) % 256;
      console.log(`[${this.pluginId}] Periodic tick — state:`, this._state);
      this.publishState(DEVICE_ID, { ...this._state });
    }, 5000);
  }

  onCommand(deviceId, payload) {
    console.log(`[${this.pluginId}] Command for ${deviceId}:`, payload);
    // Merge only known attributes to prevent arbitrary state pollution.
    for (const key of Object.keys(CAPABILITIES)) {
      if (key in payload) this._state[key] = payload[key];
    }
    this.publishState(deviceId, { ...this._state });
    console.log(`[${this.pluginId}] State after command:`, this._state);
  }
}

// ---------------------------------------------------------------------------
// CLI arg parsing (--key value)
// ---------------------------------------------------------------------------
function arg(name) {
  const i = process.argv.indexOf(`--${name}`);
  return i !== -1 ? process.argv[i + 1] : undefined;
}

const plugin = new VirtualLightPlugin({
  id:         arg('id'),
  brokerHost: arg('broker'),
  brokerPort: arg('port') ? parseInt(arg('port'), 10) : undefined,
  password:   arg('password'),
});

console.log('Virtual light plugin starting');
console.log(`  Device ID : ${DEVICE_ID}`);
console.log(`  Broker    : ${plugin.brokerHost}:${plugin.brokerPort}`);
console.log('Press Ctrl-C to stop');

const client = plugin.run();

// Publish offline on graceful shutdown.
process.on('SIGINT', () => {
  console.log('\nShutting down…');
  plugin.publishAvailability(DEVICE_ID, false);
  plugin.publishPluginStatus('offline');
  setTimeout(() => { client.end(true); process.exit(0); }, 300);
});

File: plugins/plugin-sdk-py/examples/virtual_light.py
#!/usr/bin/env python3
"""virtual_light.py — A software-only dimmable light plugin for HomeCore.

Simulates a dimmable light bulb.  On startup it registers with the broker,
publishes its initial state, then toggles on/off and steps brightness every
5 seconds.  It also responds to commands received via MQTT.

Usage::

    pip install homecore-plugin-sdk
    python virtual_light.py
    python virtual_light.py --broker 192.168.1.10 --port 1883 --id plugin.mylight
    HC_BROKER_HOST=192.168.1.10 python virtual_light.py
"""

import argparse
import logging
import threading
import time

from homecore_plugin_sdk import PluginBase

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)-8s %(name)s — %(message)s",
)
logger = logging.getLogger("virtual_light")

DEVICE_ID = "light.virtual_py_01"
CAPABILITIES = {
    "on":         {"type": "boolean"},
    "brightness": {"type": "integer", "minimum": 0, "maximum": 255},
}


class VirtualLightPlugin(PluginBase):
    """Simulated dimmable light bulb plugin."""

    PLUGIN_ID = "plugin.virtual_py"

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._state = {"on": False, "brightness": 128}
        self._lock = threading.Lock()

    # ------------------------------------------------------------------
    # PluginBase hooks
    # ------------------------------------------------------------------

    def on_connect(self) -> None:
        self.register_device(DEVICE_ID, "Virtual Light (Python)", CAPABILITIES, area="living_room")
        self.publish_availability(DEVICE_ID, True)

        with self._lock:
            state = dict(self._state)
        self.publish_state(DEVICE_ID, state)
        self.publish_plugin_status("active")

        logger.info("Device registered and initial state published: %s", state)

        # Periodic toggle in a background daemon thread so run() can block.
        t = threading.Thread(target=self._tick_loop, daemon=True, name="virtual-light-tick")
        t.start()

    def on_command(self, device_id: str, payload: dict) -> None:
        logger.info("Command received for %s: %s", device_id, payload)
        with self._lock:
            self._state.update({k: v for k, v in payload.items() if k in CAPABILITIES})
            state = dict(self._state)
        self.publish_state(device_id, state)
        logger.info("State after command: %s", state)

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _tick_loop(self) -> None:
        """Toggle on/off and step brightness every 5 seconds."""
        while True:
            time.sleep(5)
            with self._lock:
                self._state["on"] = not self._state["on"]
                self._state["brightness"] = (self._state["brightness"] + 16) % 256
                state = dict(self._state)
            logger.info("Periodic tick — publishing state: %s", state)
            self.publish_state(DEVICE_ID, state)


def main() -> None:
    parser = argparse.ArgumentParser(description="HomeCore virtual light plugin (Python)")
    parser.add_argument("--broker",   default=None, help="MQTT broker host (default: HC_BROKER_HOST or 127.0.0.1)")
    parser.add_argument("--port",     type=int, default=None, help="MQTT broker port (default: HC_BROKER_PORT or 1883)")
    parser.add_argument("--id",       default=None, help="Plugin ID override")
    parser.add_argument("--password", default=None, help="MQTT password (default: HC_PLUGIN_PASSWORD)")
    args = parser.parse_args()

    plugin = VirtualLightPlugin(
        broker_host=args.broker,
        broker_port=args.port,
        password=args.password,
    )
    if args.id:
        plugin.PLUGIN_ID = args.id

    logger.info("Virtual light plugin starting")
    logger.info("  Device ID : %s", DEVICE_ID)
    logger.info("  Broker    : %s:%s", plugin.broker_host, plugin.broker_port)
    logger.info("Press Ctrl-C to stop")

    try:
        plugin.run()
    except KeyboardInterrupt:
        logger.info("Shutting down")
        plugin.publish_availability(DEVICE_ID, False)
        plugin.publish_plugin_status("offline")


if __name__ == "__main__":
    main()
